### ISSUE TITLE
Phase 41.4 — InputSanitizer: Feature Squeezing, Denoising & Adversarial Detection

### ISSUE THREAD

**Issue by web3guru888** [web3guru888/asi-build#828]
# Phase 41.4 — InputSanitizer

## Overview

The **InputSanitizer** implements reactive defenses that transform or filter inputs before they reach the model — detecting and neutralizing adversarial perturbations at the input level. It implements feature squeezing (Xu et al., 2018), JPEG compression defense, spatial smoothing, denoising autoencoders, and anomaly-based detection — providing a first line of defense that is model-agnostic.

## Architecture

```
┌─────────────────────────────────────────────────────────────────┐
│                        InputSanitizer                            │
│                                                                  │
│  ┌─────────────┐  ┌─────────────┐  ┌──────────────────────┐    │
│  │  Detection   │  │ Sanitization│  │  Anomaly Detection   │    │
│  │  Pipeline    │  │ Pipeline    │  │                      │    │
│  │              │  │             │  │  Statistical tests    │    │
│  │  Feature     │  │  Bit-depth  │  │  Feature distance     │    │
│  │  Squeezing   │  │  Reduction  │  │  Logit analysis       │    │
│  │  Detection   │  │             │  │  Input reconstruction │    │
│  │              │  │  Spatial    │  │  Ensemble agreement    │    │
│  │  Ensemble    │  │  Smoothing  │  │                      │    │
│  │  Disagreement│  │             │  │                      │    │
│  │              │  │  JPEG       │  │                      │    │
│  │  MagNet      │  │  Compress   │  │                      │    │
│  │  Detection   │  │             │  │                      │    │
│  │              │  │  Denoising  │  │                      │    │
│  │              │  │  Autoencoder│  │                      │    │
│  └──────┬───────┘  └──────┬──────┘  └──────────┬───────────┘    │
│         │                 │                     │                │
│  ┌──────▼─────────────────▼─────────────────────▼─────────────┐  │
│  │              Decision Engine                                │  │
│  │  Accept (clean) | Sanitize (suspected) | Reject (adversarial)│  │
│  └─────────────────────────────────────────────────────────────┘  │
└─────────────────────────────────────────────────────────────────┘
```

## TypeScript Interface

```typescript
import { Tensor, Model } from '@asi-build/core';

type SanitizationMethod = 'bit_depth' | 'spatial_smooth' | 'jpeg_compress' | 
                          'denoise_ae' | 'pixel_deflection' | 'randomized_resize';

type DetectionMethod = 'feature_squeezing' | 'magnet' | 'lid' | 
                       'ensemble_disagreement' | 'input_reconstruction';

interface SanitizationConfig {
  methods: SanitizationMethod[];   // pipeline of sanitization methods
  bitDepth?: number;               // for bit-depth reduction (default 4)
  smoothingKernel?: number;        // spatial smoothing window (default 3)
  jpegQuality?: number;            // JPEG quality factor (default 75)
  denoiserModelPath?: string;      // path to denoising autoencoder
  deflectionWindow?: number;       // pixel deflection window
  resizeFactor?: number;           // randomized resize factor
}

interface DetectionConfig {
  methods: DetectionMethod[];
  threshold: number;               // detection threshold
  featureSqueezing: {
    squeezers: SanitizationMethod[];
    l1_threshold: number;          // L1 distance threshold
  };
  ensemble: {
    models: Model[];
    disagreement_threshold: number;
  };
  reconstruction: {
    autoencoder: Model;
    mse_threshold: number;
  };
}

interface SanitizationResult {
  sanitized: Tensor;
  isAdversarial: boolean;
  confidence: number;              // confidence that input is adversarial
  detectionScores: Map<string, number>;
  sanitizationApplied: SanitizationMethod[];
  processingTimeMs: number;
}

interface DetectionReport {
  totalSamples: number;
  truePositives: number;          // adversarial correctly detected
  falsePositives: number;         // clean incorrectly flagged
  trueNegatives: number;         // clean correctly passed
  falseNegatives: number;        // adversarial missed
  auroc: number;                  // area under ROC curve
  auprc: number;                  // area under precision-recall curve
  fpr95: number;                  // false positive rate at 95% true positive rate
}

interface InputSanitizer {
  // Sanitization methods
  sanitize(input: Tensor, config: SanitizationConfig): Promise<SanitizationResult>;
  bitDepthReduce(input: Tensor, bits: number): Tensor;
  spatialSmooth(input: Tensor, kernelSize: number): Tensor;
  jpegCompress(input: Tensor, quality: number): Tensor;
  denoise(input: Tensor, denoiser: Model): Promise<Tensor>;
  
  // Detection methods
  detect(input: Tensor, model: Model, config: DetectionConfig): Promise<{isAdversarial: boolean, confidence: number}>;
  featureSqueezingDetect(input: Tensor, model: Model, squeezers: SanitizationMethod[]): Promise<boolean>;
  
  // Evaluation
  evaluateDetection(cleanData: Tensor[], advData: Tensor[], model: Model, config: DetectionConfig): Promise<DetectionReport>;
  
  // Pipeline
  createPipeline(config: SanitizationConfig & DetectionConfig): (input: Tensor) => Promise<SanitizationResult>;
}
```

## Feature Squeezing Detection

```python
class FeatureSqueezingDetector:
    """Xu et al. (2018) - Feature Squeezing Detection."""
    
    def __init__(self, model, squeezers: List[Callable], threshold: float = 0.05):
        self.model = model
        self.squeezers = squeezers
        self.threshold = threshold
    
    def detect(self, x: Tensor) -> Tuple[bool, float]:
        """Detect adversarial input via prediction disagreement under squeezing."""
        original_pred = F.softmax(self.model(x), dim=1)
        
        max_distance = 0.0
        for squeezer in self.squeezers:
            squeezed_x = squeezer(x)
            squeezed_pred = F.softmax(self.model(squeezed_x), dim=1)
            distance = torch.norm(original_pred - squeezed_pred, p=1).item()
            max_distance = max(max_distance, distance)
        
        is_adversarial = max_distance > self.threshold
        return is_adversarial, max_distance

def bit_depth_reduce(x: Tensor, bits: int = 4) -> Tensor:
    """Reduce color bit depth."""
    max_val = 2 ** bits - 1
    return torch.round(x * max_val) / max_val

def spatial_smooth(x: Tensor, kernel_size: int = 3) -> Tensor:
    """Apply median filtering."""
    from torchvision.transforms.functional import gaussian_blur
    return gaussian_blur(x, kernel_size=[kernel_size, kernel_size])

def jpeg_compress(x: Tensor, quality: int = 75) -> Tensor:
    """JPEG compression and decompression."""
    from PIL import Image
    import io
    # Convert to PIL, compress, decompress
    img = to_pil(x)
    buffer = io.BytesIO()
    img.save(buffer, format='JPEG', quality=quality)
    buffer.seek(0)
    compressed = Image.open(buffer)
    return to_tensor(compressed)
```

## Configuration

```yaml
input_sanitizer:
  detection:
    enabled: true
    methods: ["feature_squeezing", "ensemble_disagreement"]
    
    feature_squeezing:
      squeezers: ["bit_depth", "spatial_smooth", "jpeg_compress"]
      bit_depth: 5
      smooth_kernel: 3
      jpeg_quality: 75
      l1_threshold: 0.05
    
    ensemble:
      disagreement_threshold: 0.3
      min_models: 3
    
    reconstruction:
      autoencoder_path: "models/denoiser.pt"
      mse_threshold: 0.01
  
  sanitization:
    pipeline: ["jpeg_compress", "spatial_smooth"]
    jpeg_quality: 80
    smooth_kernel: 3
    bit_depth: 6
    
  policy:
    action_on_detection: "sanitize"   # reject | sanitize | flag
    confidence_threshold: 0.7
    log_detections: true
    alert_on_burst: true              # alert if >10% inputs flagged
    burst_window_seconds: 60
```

## Testing Strategy

```typescript
describe('InputSanitizer', () => {
  describe('Feature Squeezing Detection', () => {
    it('should detect PGD adversarial examples', async () => {
      const advInput = await attacker.pgd(model, cleanInput, label, pgdConfig);
      const result = await sanitizer.detect(advInput.adversarial, model, detectionConfig);
      expect(result.isAdversarial).toBe(true);
    });
    
    it('should not flag clean inputs', async () => {
      const result = await sanitizer.detect(cleanInput, model, detectionConfig);
      expect(result.isAdversarial).toBe(false);
    });
    
    it('should achieve <5% false positive rate', async () => {
      const report = await sanitizer.evaluateDetection(cleanSet, advSet, model, config);
      expect(report.falsePositives / (report.falsePositives + report.trueNegatives)).toBeLessThan(0.05);
    });
  });
  
  describe('Sanitization Pipeline', () => {
    it('should preserve clean accuracy after sanitization', async () => {
      const sanitized = await sanitizer.sanitize(cleanInput, sanitizeConfig);
      const pred = model(sanitized.sanitized);
      expect(pred.argmax()).toBe(label);
    });
    
    it('should reduce adversarial success rate', async () => {
      const sanitized = await sanitizer.sanitize(advInput, sanitizeConfig);
      const pred = model(sanitized.sanitized);
      expect(pred.argmax()).toBe(originalLabel);
    });
  });
  
  describe('JPEG Compression', () => {
    it('should be differentiable-attack aware (Athalye et al.)', async () => {
      // Verify defense is not trivially bypassed by BPDA
      const bpdaAttack = await attacker.pgd(model, cleanInput, label, { ...pgdConfig, useEOT: true });
      const sanitized = await sanitizer.jpegCompress(bpdaAttack.adversarial, 75);
      // At least partially effective even under adaptive attack
    });
  });
});
```

## Acceptance Criteria

- [ ] Feature squeezing detection achieves ≥90% AUROC on PGD adversarial examples
- [ ] False positive rate on clean data <5% at detection threshold
- [ ] JPEG compression defense reduces attack success rate by ≥30%
- [ ] Sanitization pipeline preserves ≥95% clean accuracy
- [ ] Denoising autoencoder reconstructs clean inputs with MSE <0.01
- [ ] Detection pipeline processes inputs in <10ms (real-time capable)
- [ ] Burst detection alerts when adversarial rate exceeds threshold
- [ ] Evaluation includes adaptive attack analysis (Athalye et al., 2018; Tramèr et al., 2020)

## References

- Xu et al. (2018) — Feature Squeezing: Detecting Adversarial Examples in DNNs
- Dziugaite et al. (2016) — A Study of the Effect of JPG Compression on Adversarial Images
- Meng & Chen (2017) — MagNet: A Two-Pronged Defense Against Adversarial Examples
- Athalye et al. (2018) — Obfuscated Gradients Give a False Sense of Security
- Tramèr et al. (2020) — On Adaptive Attacks to Adversarial Example Defenses

---
*Sub-phase of [Phase 41 — Adversarial Robustness & Security Intelligence](../discussions/824)*
